File: envdiff/parsers/compose.py
from __future__ import annotations

import re
from pathlib import Path

from envdiff.models import EnvVarUsage, UsageScanResult
from envdiff.utils.ordering import sort_usages

INTERPOLATION_RE = re.compile(r"\$\{([A-Z0-9_]+)(?::-(.*?))?\}")


def scan_compose_file(path: str | Path) -> UsageScanResult:
    file_path = Path(path)
    usages: list[EnvVarUsage] = []

    for line_number, line in enumerate(file_path.read_text(encoding="utf-8").splitlines(), start=1):
        for match in INTERPOLATION_RE.finditer(line):
            default_value = match.group(2)
            usages.append(
                EnvVarUsage(
                    name=match.group(1),
                    file_path=str(file_path),
                    line_number=line_number,
                    usage_kind="compose_interpolation",
                    requiredness=(
                        "optional_with_default" if default_value is not None else "required"
                    ),
                    default_value=default_value,
                    source_type="compose",
                )
            )

    return UsageScanResult(usages=sort_usages(usages), warnings=())

File: internal/parsers/compose.go
package parsers

import (
	"regexp"

	"github.com/davisbuilds/envdiff/internal/model"
	"github.com/davisbuilds/envdiff/internal/order"
)

var composeInterpolationPattern = regexp.MustCompile(`\$\{([A-Z0-9_]+)(:-(.*?))?\}`)

func ScanComposeFile(path string) (model.UsageScanResult, error) {
	lines, err := readLines(path)
	if err != nil {
		return model.UsageScanResult{}, err
	}

	usages := []model.EnvVarUsage{}
	for index, line := range lines {
		lineNumber := index + 1
		for _, match := range composeInterpolationPattern.FindAllStringSubmatchIndex(line, -1) {
			name := line[match[2]:match[3]]
			var defaultValue *string
			requiredness := "required"
			if match[4] != -1 {
				value := line[match[6]:match[7]]
				defaultValue = &value
				requiredness = "optional_with_default"
			}
			usages = append(usages, model.EnvVarUsage{
				DefaultValue: defaultValue,
				FilePath:     path,
				LineNumber:   model.IntPtr(lineNumber),
				Name:         name,
				Requiredness: requiredness,
				SourceType:   "compose",
				UsageKind:    "compose_interpolation",
			})
		}
	}

	return model.UsageScanResult{Usages: order.Usages(usages), Warnings: []string{}}, nil
}
